modules/core/src/smithy4s/internals/DocumentEncoderSchemaVisitor.scala

```diff
@@ -45,33 +45,44 @@ import smithy4s.schema.Primitive.PUUID
 import smithy4s.schema.Primitive.PDouble
 import smithy4s.schema.Primitive.PLong
 import smithy4s.schema.Primitive.PString
+import smithy4s.api.Untagged
 
 trait DocumentEncoder[A] { self =>
 
   def canBeKey: Boolean
 
   def apply: A => Document
-  def contramap[B](f: B => A): DocumentEncoder[B] =
+  final def contramap[B](f: B => A): DocumentEncoder[B] =
     new DocumentEncoder[B] {
-
       def apply: B => Document = f andThen self.apply
 
       def canBeKey: Boolean = self.canBeKey
     }
 
+  final def mapDocument(f: Document => Document): DocumentEncoder[A] =
+    new DocumentEncoder[A] {
+      def apply: A => Document = self.apply andThen f
+      def canBeKey: Boolean = self.canBeKey
+    }
+
 }
 
 object DocumentEncoder {
 
-  implicit val contraInstance: Contravariant[DocumentEncoder] =
-    new Contravariant[DocumentEncoder] {
-      def contramap[A, B](fa: DocumentEncoder[A])(
-          f: B => A
-      ): DocumentEncoder[B] = fa.contramap(f)
+  implicit val encoderKInstance: EncoderK[DocumentEncoder, Document] =
+    new EncoderK[DocumentEncoder, Document] {
+      def apply[A](fa: DocumentEncoder[A], a: A): Document = fa.apply(a)
+      def absorb[A](f: A => Document): DocumentEncoder[A] =
+        new DocumentEncoder[A] {
+          def apply: A => Document = f
+          override def canBeKey: Boolean = false
```

Comment:
this is cheating a little but we're planning on removing that `canBeKey` stuff in a subsequent PR